icu4c/source/i18n/collationdatabuilder.cpp

```diff
@@ -1391,7 +1419,71 @@ CollationDataBuilder::buildMappings(CollationData &data, UErrorCode &errorCode)
     setDigitTags(errorCode);
     setLeadSurrogates(errorCode);
 
-    if (!icu4xMode) {
+    if (icu4xMode) {
+        // Make sure that starters that occur is the middle of a
+        // contraction have contraction ce32 with the
+        // `CONTRACT_HAS_STARTER` flag set so that starters that
+        // can occur in a non-final position in a contraction can
+        // be easily recognized from having a contraction ce32
+        // that has the `CONTRACT_HAS_STARTER` flag set.
+
+        UCharsTrieBuilder contractionBuilder(errorCode);
+        // Intentionally unpaired low surrogate to make it never
+        // match well-formed UTF-16 which ICU4X feeds to the
+        // matcher.
+        UnicodeString placeholder(0xDC00);
+
+        UnicodeSetIterator iter(contractionMiddleStarter);
+        while(iter.next() && !iter.isString()) {
+            UChar32 c = iter.getCodepoint();
+            uint32_t ce32 = utrie2_get32(trie, c);
+            UBool fromBase = false;
+            if(ce32 == Collation::FALLBACK_CE32) {
+                fromBase = true;
+                ce32 = base->getCE32(c);
+            }
+            if (!(Collation::hasCE32Tag(ce32, Collation::CONTRACTION_TAG) && (ce32 & Collation::CONTRACT_HAS_STARTER))) {
+                if (fromBase) {
+                    // This case does not actually happen as of February 2025.
+                    ce32 = copyFromBaseCE32(c, ce32, true, errorCode);
+                }
+                if (Collation::hasCE32Tag(ce32, Collation::CONTRACTION_TAG)) {
+                    // This middle starter is also the first character of another
+                    // contraction, but that contraction does not have the
+                    // CONTRACT_HAS_STARTER flag. Let's add the flag to make
+                    // mark this at the expense of pessimizing the matching
+                    // of this contraction.
+                    // As of February 2025, this case does not actually occur
+                    // in CLDR.
+                    ce32 |= Collation::CONTRACT_HAS_STARTER;
+                } else {
+                    // This middle starter is not also the first character
+                    // in another contraction.
+
+                    // The UCharsTrie needs to contain some placeholder. Trying to
+                    // build an empty UCharsTrie causes an error. We build a trie
```

Comment:
```suggestion
                    // The UCharsTrie needs to contain some placeholder
                    // because it cannot be empty. We build a trie
```